portal/models/qb_timeline.py

```diff
@@ -315,11 +316,131 @@ def qbds_for_rp(rp, classification, trigger_date):
                 retired=next_retired,
                 qbds=qbds_for_rp(next_rp, classification, trigger_date)
             )
+            if curRPD.retired == nextRPD.retired:
+                raise ValueError(
+                    "Invalid state: multiple RPs w/ same retire date")
         else:
             nextRPD = None
         yield curRPD, nextRPD
 
 
+class RP_flyweight(object):
+    """maintains state for RPs as it transitions
+
+    Any number of Research Protocols may apply to a user.  This
+    class manages transitioning through the applicable as time
+    passes, maintaining a "current" and "next" as well as the ability
+    to continue marching forward.
+
+    """
+
+    def __init__(self, user, trigger_date, classification):
+        """Initialize flyweight state
+        :param user: the patient
+        :param trigger_date: the patient's initial trigger date
+        :param classification: `indefinite` or None
+        """
+        self.user = user
+        self.td = trigger_date
+        self.classification = classification
+        self.rp_walker = cur_next_rp_gen(
+            user=self.user, trigger_date=self.td,
+            classification=self.classification)
+        self.cur_rpd, self.nxt_rpd = next(self.rp_walker, (None, None))
+        self.skipped_nxt_start = None
+
+    def adjust_start(self):
+        """The QB start may need a minor adjustment, done once when ready"""
+        self.cur_qbd.relative_start = self.cur_start
+
+        # sanity check - make sure we don't adjust twice
+        if hasattr(self.cur_qbd, 'already_adjusted'):
+            raise RuntimeError(
+                'already adjusted the qbd relative start')
+        self.cur_qbd.already_adjusted = True
+
+    def consider_transition(self):
+        """Returns true only if state suggests it *may* be transtion time"""
+        return self.nxt_rpd and self.cur_rpd.retired < self.cur_exp
+
+    def next_qbd(self):
+        """Advance to next qbd on applicable RPs"""
+        self.cur_qbd, self.cur_start, self.cur_exp = None, None, None
+        if self.cur_rpd:
+            self.cur_qbd = next(self.cur_rpd.qbds, None)
+        if self.cur_qbd:
+            self.cur_start = calc_and_adjust_start(
+                user=self.user, qbd=self.cur_qbd, initial_trigger=self.td)
+            self.cur_exp = calc_and_adjust_expired(
+                user=self.user, qbd=self.cur_qbd, initial_trigger=self.td)
+
+        self.nxt_qbd, self.nxt_start = None, None
+        if self.nxt_rpd:
+            self.nxt_qbd = next(self.nxt_rpd.qbds, None)
+        if self.nxt_qbd:
```

Comment:
might be good to invert into a guard clause, to remove some of the below nesting

Reply:
yes, looks like i'm too late... already merged :)